SingularityS3Downloader/src/main/java/com/hubspot/singularity/s3downloader/server/SingularityS3DownloaderAsyncHandler.java

```diff
@@ -19,33 +25,112 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(SingularityS3DownloaderAsyncHandler.class);
 
+  private final long start;
+
   private final ArtifactDownloadRequest artifactDownloadRequest;
-  private final Continuation continuation;
   private final ArtifactManager artifactManager;
-  private final long start;
   private final SingularityS3DownloaderMetrics metrics;
 
+  private final AtomicBoolean finished;
+  private final Lock continuationsLock;
+  private final List<Continuation> continuations;
+
   public SingularityS3DownloaderAsyncHandler(ArtifactManager artifactManager, ArtifactDownloadRequest artifactDownloadRequest, Continuation continuation, SingularityS3DownloaderMetrics metrics) {
     this.artifactManager = artifactManager;
     this.artifactDownloadRequest = artifactDownloadRequest;
-    this.continuation = continuation;
     this.metrics = metrics;
     this.start = System.currentTimeMillis();
+
+    this.finished = new AtomicBoolean(false);
+    this.continuationsLock = new ReentrantLock();
+    this.continuations = Lists.newArrayList(continuation);
+  }
+
+  public boolean addContinuation(Continuation continuation) {
+    if (finished.get()) {
+      return false;
+    }
+
+    try {
+      continuationsLock.lock();
+
+      if (finished.get()) {
+        return false;
+      }
+
+      continuations.add(continuation);
+    } finally {
+      continuationsLock.unlock();
+    }
+
+    return true;
+  }
+
+  private boolean isExpired() {
+    try {
+      continuationsLock.lock();
+
+      boolean isExpired = true;
+
+      for (Continuation continuation : continuations) {
+        if (!continuation.isExpired()) {
+          isExpired = false;
+        }
+      }
+
+      if (isExpired) {
+        finished.set(true);
+      }
+
+      return isExpired;
+    } finally {
+      continuationsLock.unlock();
+    }
+  }
+
+  private void closeContinuations() {
+    for (Continuation continuation : continuations) {
+      try {
+        getResponse(continuation).getOutputStream().close();
+      } catch (IOException e) {
+        LOG.warn("Error closing continuation for {}", artifactDownloadRequest, e);
+      }
+    }
+  }
+
+  private HttpServletResponse getResponse(Continuation continuation) {
+    return (HttpServletResponse) continuation.getServletResponse();
+  }
+
+  private void completeContinuations() {
+    for (Continuation continuation : continuations) {
+      continuation.complete();
+    }
+  }
+
+  private void sendErrors() {
+    for (Continuation continuation : continuations) {
+      try {
+        getResponse(continuation).getOutputStream().close();
+      } catch (IOException e) {
+        LOG.warn("Error sending error for {}", artifactDownloadRequest, e);
+      }
+    }
```

Comment:
This method looks like same as `closeContinuations` (minus the different log message)